app/services/form_response.rb

```diff
@@ -1,8 +1,11 @@
 class FormResponse
-  def initialize(success:, errors:, extra: {})
+  def initialize(success:, errors: {}, extra: {})
     @success = success
-    @errors = errors.to_hash
+    @errors = errors.is_a?(ActiveModel::Errors) ? errors.messages.to_hash : errors
     @extra = extra
+    @extra.merge!(
+      error_details: flatten_details(errors.details),
```

Comment:
1. what if we made `error_details` an attribute, and only put it in the hash at the end? then we could simplify `merge` because we wouldn't need to special-case the `error_details` key and could just append the two arrays in that method?
2. Do we need to a check that errors has `#details` before calling this? Because plain hashes will not have `#details`

Reply:
> 1. what if we made `error_details` an attribute, and only put it in the hash at the end? then we could simplify `merge` because we wouldn't need to special-case the `error_details` key and could just append the two arrays in that method?

The original motivation for this approach was to avoid making those details part of the public interface, where nobody should ever need to explicitly pass `error_details`, instead relying on it to be handled internally when given an `ActiveModel::Errors` instance. Though I suppose if this were `attr_accessor`, we wouldn't necessarily need to make this part of the initialize signature, but could still handle it in the merge? I'll poke at it a bit more.

> 2. Do we need to a check that errors has `#details` before calling this? Because plain hashes will not have `#details`

The condition on the next line should avoid any issues here?

https://github.com/18F/identity-idp/blob/445c4ad6c45bbcc8a270eeacaf0b97a619b5c817/app/services/form_response.rb#L8

Maybe there's some weird cases in `merge` if someone manually defines an `error_details` property of a hash, but the problem might go away depending on changes regarding your first point.

Reply:
> The condition on the next line should avoid any issues here?

🤦 somehow didn't see that line ya that works

Reply:
Maybe some `instance_eval` to get around the property access? It's ugly but gets the job done... 😬 

```ruby
class A
  attr_accessor :a

  def initialize(a)
    @a = a
    @b = a + 1
  end

  def merge(other)
    self.class.new(self.a + other.a).tap do |merged|
      outer_b = b
      merged.instance_eval { @b = outer_b + other.instance_eval { b } }
    end
  end

  private

  attr_accessor :b
end

puts A.new(1).merge(A.new(2))
```

Reply:
That could work 😄  947e9e6c2def547b42e82025b48a3d16519aa54c